### DISCUSSION TITLE
How do I get migrations to work when testing within a Package (not an Application)

### DISCUSSION THREAD

**Discussion by mv12ks** [spatie/laravel-permission#2478] [Q&A] [answered]
I have a Laravel package that uses spatie/laravel-permission. I want to test my permissions. I copied the [migration stub](https://github.com/spatie/laravel-permission/blob/main/database/migrations/create_permission_tables.php.stub) to the migrations folder of my package to create the tables.

Laravel 10.
PHP 8.2
Package: 5.10

In my tests, I want to run my seeder that inserts some roles and permissions into the tables. The service provider of my package registers the migrations like so:

```php
public function register() : void
{
    $this->loadMigrationsFrom(__DIR__ . '/../database/migrations');
}
```

My test case looks like this:

```php
class PermissionsTest extends TestCase
{
    use RefreshDatabase;

    /** @testdox .... */
    public function test_some_permission() : void
    {
        $this->seed(RolesAndPermissionsSeeder::class);

        // ... assertions
    }

    /**
     * Prepare the testing environment.
     *
     * @param \Illuminate\Foundation\Application $app
     * @return void
     */
    protected function getEnvironmentSetUp($app) : void
    {
        $app['config']->set('database.default', 'testing');
        $app['config']->set('database.connections.testing', [
            'driver'   => 'sqlite',
            'database' => ':memory:',
            'prefix'   => '',
        ]);
    }

    /**
     * The migrations to load prior to testing.
     *
     * @return void
     */
    protected function defineDatabaseMigrations() : void
    {
        $this->loadLaravelMigrations();
    }
}
```

This test does not run. As soon as I comment out the loadMigrationsFrom method, the error dissapears. I think it might be because the migration uses `config()`, but I am not getting an error message, not even when using `$this->withoutExceptionHandling();`.

Any tips as to where to start debugging this, or any fix?

**Comment by parallels999** [comment]
Laravel version? package version? php version?

**Reply by mv12ks** [reply]
Laravel 10.
PHP 8.2
Package: 5.10

**Reply by parallels999** [reply]
Maybe
https://github.com/spatie/laravel-permission/discussions/1676#discussioncomment-398560
https://github.com/spatie/laravel-permission/discussions/1939#discussioncomment-1728504

**Reply by mv12ks** [reply]
These do not work, I've tested them just now.

**Comment by mv12ks** [comment]
It is because the config is not present because it is a package.

I can do this in my test case OR service provider and it fixes it:

```
private function setLaravelPermissionSettings() : void
{
    config()->set('permission.teams', false);

    config()->set('permission.models', [
        'permission' => \Spatie\Permission\Models\Permission::class,
        'role' => \Spatie\Permission\Models\Role::class,
    ]);

    config()->set('permission.table_names', [
        'roles' => 'roles',
        'permissions' => 'permissions',
        'model_has_roles' => 'model_has_roles',
        'model_has_permissions' => 'model_has_permissions',
        'role_has_permissions' => 'role_has_permissions',
    ]);

    config()->set('permission.column_names', [
        'role_pivot_key' => null,
        'permission_pivot_key' => null,
        'model_morph_key' => 'model_id',
        'team_foreign_key' => 'team_id',
    ]);
}
```

This works, but it is really hacky. Anyone with a better solution?

**Comment by erikn69** [comment, marked as answer]
Did you try adding
```php
/**
 * @param  \Illuminate\Foundation\Application  $app
 * @return array
 */
protected function getPackageProviders($app)
{
    return [
        \Spatie\Permission\PermissionServiceProvider::class,
    ];
}
```

**Reply by mv12ks** [reply]
This worked, thanks. A lot cleaner than setting the config values manually.
